### PR TITLE
[ADP-3306] Make `NoSuchWallet` error machine-readable.

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] Extend `NoSuchWallet` to include `WalletId`.
- [x] Adjust unit tests and swagger accordingly.
- [x] use `NoSuchWallet` rather than `err403NoSuchErrorMsg` throughout integration tests   

### Issue
ADP-3306

### CODE REVIEW

File: lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 1]
```diff
@@ -2378,7 +2381,9 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
         let link = Link.getTransaction @'Shelley w (ApiTxId txid)
         r <- request @(ApiTransaction n) ctx link Default Empty
         expectResponseCode HTTP.status404 r
-        expectErrorMessage (errMsg404NoWallet $ w ^. walletId) r
+        decodeErrorInfo r `shouldBe`
+            (NoSuchWallet $ ApiErrorNoSuchWallet $ w ^. walletId)
+
```
Comment (jonathanknowles) [inline lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/Transactions.hs]:
```suggestion
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 2]
```diff
@@ -159,6 +160,8 @@ data ApiErrorInfo
         !ApiErrorNoSuchPool
     | NoSuchTransaction
     | NoSuchWallet
+        !ApiErrorNoSuchWallet
+    | NoSuchWalletInitialized
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: If this error is a **_known_** wallet that is simply not initialized, then could we use a name like:
```suggestion
    | WalletNotInitialized
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 3]
```diff
@@ -312,3 +315,10 @@ data ApiErrorNoSuchPool = ApiErrorNoSuchPool
     deriving (Data, Eq, Generic, Show, Typeable)
     deriving (FromJSON, ToJSON) via DefaultRecord ApiErrorNoSuchPool
     deriving anyclass NFData
+
+data ApiErrorNoSuchWallet = ApiErrorNoSuchWallet
+    { walletId :: !Text
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Suggestion: does it make sense to use `ApiT WalletId` here?

```suggestion
    { walletId :: !(ApiT WalletId)
```

File: lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs [hunk 4]
```diff
@@ -190,8 +195,9 @@ scenario_ADDRESS_LIST_04 fixture = it title $ \ctx -> runResourceT $ do
     r <- request @[ApiAddressWithPath n] ctx (Link.listAddresses @'Byron w) Default Empty
     verify r
         [ expectResponseCode HTTP.status404
-        , expectErrorMessage $ errMsg404NoWallet $ w ^. walletId
         ]
+    decodeErrorInfo r `Lifted.shouldBe`
+        (NoSuchWallet $ ApiErrorNoSuchWallet $ w ^. walletId)
```
Comment (jonathanknowles) [inline lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs]: For cases like this, I would suggest:

```suggestion
        NoSuchWallet (ApiErrorNoSuchWallet $ w ^. walletId)
```

File: specifications/api/swagger.yaml [hunk 5]
```diff
@@ -4361,6 +4361,18 @@ x-errNoUtxosAvailable: &errNoUtxosAvailable
       type: string
       enum: ['no_utxos_available']
 
+x-errNoSuchWalletInitialized: &errNoSuchWalletInitialized
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml]: Related to https://github.com/cardano-foundation/cardano-wallet/pull/4597#discussion_r1606488407 (above):

```suggestion
x-errWalletNotInitialized: &errWalletNotInitialized
```

File: specifications/api/swagger.yaml [hunk 6]
```diff
@@ -4361,6 +4361,18 @@ x-errNoUtxosAvailable: &errNoUtxosAvailable
...
+  <<: *responsesErr
+  title: no_such_wallet_initialized
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml]: Related to https://github.com/cardano-foundation/cardano-wallet/pull/4597#discussion_r1606488407 (above):

```suggestion
  title: wallet_not_initialized
```

File: specifications/api/swagger.yaml [hunk 7]
```diff
@@ -5627,6 +5645,13 @@ x-responsesErr404WalletNotFound: &responsesErr404WalletNotFound
       application/json:
         schema: *errNoSuchWallet
 
+x-responsesErr404DbNotInitialized: &responsesErr404DbNotInitialized
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml]: Related to https://github.com/cardano-foundation/cardano-wallet/pull/4597#discussion_r1606488407 (above):

How about:
```suggestion
x-responsesErr404WalletNotInitialized: &responsesErr404WalletNotInitialized
```

File: specifications/api/swagger.yaml [hunk 8]
```diff
@@ -5627,6 +5645,13 @@ x-responsesErr404WalletNotFound: &responsesErr404WalletNotFound
...
+  404:
+    description: DB uninitialized
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml]:
```suggestion
    description: Wallet not yet initialized.
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @paweljakubas 

Thanks for making this PR! I've left a few suggestions.

I believe the unit test failure will be fixed by: https://github.com/cardano-foundation/cardano-wallet/pull/4598.

One last request: would it be possible to fill out the PR description?

Many thanks again!

File: lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs (hunk 4 above)
Comment (jonathanknowles) [reply lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs]: (I think `hlint` highlights a series of similar cases.)

File: lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/Transactions.hs]: fixed

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: sounds reasonable - done

File: specifications/api/swagger.yaml (hunk 5 above)
Comment (paweljakubas) [reply specifications/api/swagger.yaml]: done

File: lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs]: indeed, done

Comment (jonathanknowles) [reply lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs]: @paweljakubas `hlint` is also failing for these: https://buildkite.com/cardano-foundation/cardano-wallet/builds/4956#018f8702-7a5f-465b-8b49-1b376461a17a

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: trying to use pattern synonyms here -> https://github.com/cardano-foundation/cardano-wallet/pull/4597/commits/984bb6eb3c841acaefb22251c42f58a0b7327462

**Review by jonathanknowles — APPROVED** [review]
LGTM!

@paweljakubas, many thanks for making this PR!

**Pull request merged**